### PR TITLE
Snapshot tests & Test framework improvements

### PR DESCRIPTION
**Pull request by dkijania**
Added snapshot tests:

1. Functional
- big snapshot 100k wallets import
- replacing existing one or adding new tags

2. Load:
- load test for already imported snapshot
- updating snapshot during load test

During initial development I've also noticed that rest client module could be improved and RawRestClient (for operations without need of serialization, like expecting error code, or checking just status) can be extracted. Updated also usages. Finally cleaned up non_functional modulel

### CODE REVIEW

File: vit-servicing-station-tests/src/common/clients/rest/raw.rs [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+        let mut res = self.client()?.post(path).body(String::into_bytes(data));
+
+        if let Some(api_token) = &self.api_token {
+            res = res.header(API_TOKEN_HEADER, api_token.to_string());
+        }
+        let response = res.send()?;
+        Ok(response)
+    }
+
+    pub fn get(&self, path: &str) -> Result<reqwest::blocking::Response, reqwest::Error> {
```
Comment (enzoc4) [inline vit-servicing-station-tests/src/common/clients/rest/raw.rs:151]: maybe this shouldn't be `pub`

File: vit-servicing-station-tests/src/common/clients/rest/raw.rs [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+        }
+        if let Some(origin) = &self.origin {
+            res = res.header(ORIGIN, origin.to_string());
+        }
+        let response = res.send()?;
+        self.logger.log_response(&response);
+        Ok(response)
+    }
+
+    pub fn put(
```
Comment (enzoc4) [inline vit-servicing-station-tests/src/common/clients/rest/raw.rs:166]: same

File: vit-servicing-station-tests/src/common/load/mod.rs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    pub fn random_token(&mut self) -> String {
+        let tokens = self.snapshot.tokens();
+        let random_idx = self.random_usize() % tokens.len();
+        tokens
+            .keys()
+            .cloned()
+            .collect::<Vec<String>>()
+            .get(random_idx)
```
Comment (enzoc4) [inline vit-servicing-station-tests/src/common/load/mod.rs:31]:
```suggestion
            .nth(random_idx)
```

File: vit-servicing-station-tests/src/common/load/rest.rs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        }
+    }
+}
+
+impl RequestGenerator for VitRestRequestGenerator {
+    fn next(&mut self) -> std::result::Result<Request, RequestFailure> {
+        self.rest_client
+            .set_api_token(self.snapshot_randomizer.random_token());
+
+        match self.snapshot_randomizer.random_usize() % 7 {
```
Comment (enzoc4) [inline vit-servicing-station-tests/src/common/load/rest.rs:29]: why 7 if there are only 3 cases handled?

File: vit-servicing-station-tests/src/common/load/voting_power.rs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        }
+    }
+}
+
+impl RequestGenerator for VotingPowerRequestGenerator {
+    fn next(&mut self) -> std::result::Result<Request, RequestFailure> {
+        let mut rng = rand::rngs::OsRng;
+        let content = &self.snapshot.content;
+        self.rest_client
+            .voting_power(
```
Comment (enzoc4) [inline vit-servicing-station-tests/src/common/load/voting_power.rs:34]: what's this request for?

File: vit-servicing-station-tests/src/common/load/voting_power.rs [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+            )
+            .map(|_| Request {
+                ids: vec![Some(rng.next_u64().to_string())],
+                duration: Duration::ZERO,
+            })
+            .map_err(|e| RequestFailure::General(format!("voting power: {}", e)))
+    }
+
+    fn split(mut self) -> (Self, Option<Self>) {
+        // Since this is idempotent calls we could split as many times as we want
```
Comment (enzoc4) [inline vit-servicing-station-tests/src/common/load/voting_power.rs:46]: what's idempotent? `split` doesn't look like it

File: vit-servicing-station-tests/src/tests/snapshot.rs [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+        );
+    }
+    for (idx, entry) in second_snapshot.content.iter().enumerate() {
+        let voting_power = VotingPower::from(entry.clone());
+        assert_eq!(
+            vec![voting_power],
+            rest_client
+                .voting_power(&second_snapshot.tag, &entry.voting_key.to_hex())
+                .unwrap(),
+            "expected empty data for entry idx: {}",
```
Comment (enzoc4) [inline vit-servicing-station-tests/src/tests/snapshot.rs:94]: non-empty data, actually

File: vit-servicing-station-tests/src/tests/snapshot.rs [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+    let rest_client = server.rest_client_with_token(&data.token_hash());
+
+    let snapshot = SnapshotBuilder::default()
+        .with_tag("big")
+        .with_entries_count(100_000)
+        .with_groups(vec![
+            "daily".to_string(),
+            "fund8".to_string(),
+            "fund9".to_string(),
+            "fund10".to_string(),
```
Comment (enzoc4) [inline vit-servicing-station-tests/src/tests/snapshot.rs:165]: these names don't really look like what I would expect as `groups`, but it's mostly irrelevant

File: vit-servicing-station-tests/src/common/snapshot.rs [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+impl From<Snapshot> for SnapshotUpdater {
+    fn from(snapshot: Snapshot) -> Self {
+        Self { snapshot }
+    }
+}
+
+impl SnapshotUpdater {
+    pub fn update_tag<S: Into<String>>(mut self, tag: S) -> Self {
```
Comment (enzoc4) [inline vit-servicing-station-tests/src/common/snapshot.rs:104]: `with_tag` sound better to me

File: vit-servicing-station-tests/src/common/load/rest.rs (hunk 4 above)
Comment (dkijania) [reply vit-servicing-station-tests/src/common/load/rest.rs:29]: good catch!

File: vit-servicing-station-tests/src/common/load/voting_power.rs (hunk 5 above)
Comment (dkijania) [reply vit-servicing-station-tests/src/common/load/voting_power.rs:34]: for voting_power and group.. maybe i should rename it?

**Pull request merged**